### PR TITLE
Copy session and auth code from other PR, add example app

### PR DESCRIPTION
**Pull request by imalsogreg**
This is largely a copy-paste from https://github.com/dzhus/snaplet-redis/pull/9, tweaked for snap >= 1.0, and with an example snap app.

I've also put in email-based user lookup, for the upcoming change to the auth snaplet.

### CODE REVIEW

**Comment by dzhus** [comment]
@imalsogreg Can you please update `package.yaml` to include new dependencies so that the builds passes?

File: src/Snap/Snaplet/Auth/Backends/Redis.hs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+data RedisAuthManager = RedisAuthManager {
+                      conn :: Connection
+                      }
+
+instance IAuthBackend RedisAuthManager where
+  save = redisSave
+  destroy = redisDestroy
+  lookupByUserId = redisLookupByUserId
+  lookupByLogin = redisLookupByLogin
+  lookupByEmail = redisLookupByEmail
```
Comment (dzhus) [inline src/Snap/Snaplet/Auth/Backends/Redis.hs]: It's `lookupByEmail` that needs to be `ifdef`'d

**Comment by imalsogreg** [comment]
@dzhus After a few tries, the new stack resolver succeeds. Willing do bump the older of the two resolvers? Snap itself only supports the past three ghc versions.. do you know what resolver that corresponds to?

**Comment by dzhus** [comment]
At the moment, `stack.yaml` is GHC 8.0 series with snap-1.0.x and `stack-previous.yaml` is GHC 7.10 (https://www.stackage.org/nightly-2016-05-25) that also uses snap-0.9.x. The PR doesn't seem to build with snap-0.9 (ifdefing needed around `setSessionCookie` due to a breaking change in https://github.com/snapframework/snap/commit/0ac4ce7035a4bba906f934d14464a3c01b9a8541#diff-0b2925102126c9acb46d205478887098R85), you can test it by doing `stack --stack-yaml=stack-previous.yaml build` - let me know if you want to fix that or I'll merge it as is.

**Comment by dzhus** [comment]
@imalsogreg If you do any further changes please rebase onto 68e7244

File: src/Snap/Snaplet/Session/Backends/RedisSession.hs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+getPayload mgr = getSecureCookie (cookieName mgr) (siteKey mgr) (timeOut mgr)
+
+
+------------------------------------------------------------------------------
+-- | Set the client-side value
+setPayload :: RedisSessionManager -> Payload -> Snap ()
+setPayload mgr = setSecureCookie
+    (cookieName mgr)
+#if MIN_VERSION_snap(1,0,0)
+    Nothing
```
Comment (dzhus) [inline src/Snap/Snaplet/Session/Backends/RedisSession.hs]: Ok, why not do the same as in https://github.com/snapframework/snap/blob/ed168734edd0b19083924d922822e9ad1314bb8a/src/Snap/Snaplet/Session/Backends/CookieSession.hs#L85 ? :)

**Comment by imalsogreg** [comment]
@dzhus Thanks for all the help. Builds are green and hlint-clean now on both resolvers. I rebased several commits, but not any in a range overlapping with merges - sounds ok, or should the commit history be made cleaner?

File: src/Snap/Snaplet/Session/Backends/RedisSession.hs (hunk 2 above)
Comment (imalsogreg) [reply src/Snap/Snaplet/Session/Backends/RedisSession.hs]: Good catch! Done.

**Comment by dzhus** [comment]
Rebased onto master and merged, thank you! I will try to push the new version to Hackage shortly.

**Comment by imalsogreg** [comment]
Nice, thanks!

**Comment by dzhus** [comment]
@imalsogreg This was finally pushed to Hackage as http://hackage.haskell.org/package/snaplet-redis-0.1.5

**Comment by imalsogreg** [comment]
Wonderful. Thanks @dzhus!

**Pull request closed without merging**